### PR TITLE
SM-462 - Sylo Action Permissions Pallet

### PR DESCRIPTION
**Pull request by JCSanPedro**
# Description

This PR introduces the new pallet: sylo-action-permissions, which allows users to grant transact permissions to other users, which then enables that permissioned user to execute transactions on behalf of the grantor.

### Features

- Allows grantees to execute extrinsics on behalf of grantor
- Allows specifying a set of allowed calls and using wildcard specifier (`*`) to enable multiple calls at once
- Allows specifying wether the spender of an extrinsic should be the GRANTOR or GRANTEE
- Enables the grantor to set a spending balance which is decremented whenever the grantee executes a call on their behalf
- Adds the `accept_transact_permission` call. This involves a grantor to signing a TransactPermissionToken, which the grantee can then submit to create the transact permission without the grantor needing to submit the token themselves.

## Context & Background

RFC: https://www.notion.so/futureverse/RFC-Sylo-Action-Permissions-Pallet-1c10cb4dab3d80bcbd09f3740fde4213

- _Add any relevant context or background information about the changes_

## Notes & Additional Information

- The pallet implementation **does not** use the proxy pallet to achieve this. This is to avoid users being able to bypass the spendor and allowed_call mechanisms by using the proxy pallet directly.
- The spendor and spending balance mechanism is achieved by modifying the [withdraw_fee](https://github.com/futureversecom/trn-seed/blob/SM-462-actions-pallet/pallet/fee-proxy/src/impls.rs#L76) implementation
- A set of e2e tests have been added at e2e/Sylo/SyloActions.tests.ts. The suite of tests attempt to validate various edge cases related futurepass calls, fee proxy calls, and xrpl calls.

---

# Release Notes

## Type of Change

- [ ] Runtime Changes
- [ ] Client Changes

#### Added

- _e.g. **Marketplace: `sellNft`**_

#### Changed

-

#### Removed

-

#### Added

- _e.g. **XRPLBridge: PaymentDelay, changed from value to map keyed by asset_id**_

#### Changed

-

#### Removed

-

#### Added

-

#### Changed

- _e.g. **XRPLBridge: set_payment_delay, added asset_id**_

#### Removed

-

#### Added

- _e.g. **XRPL-Bridge: XrplAssetMapSet**_

#### Changed

- _e.g. **XRPLBridge: PaymentDelaySet, added asset_id parameter**_

#### Removed

-

#### Added

- _e.g. **Marketplace: NoToken**_

#### Changed

-

#### Removed

-

#### Added

- _e.g. **XRPLBridge: Migrate PaymentDelay to include asset_id as key**_

#### Removed

-

---

### CODE REVIEW

File: pallet/fee-proxy/src/impls.rs [hunk 1]
```diff
@@ -90,6 +95,35 @@ where
 		// Attribute the fee to the partner if the caller is attributed to a partner
 		attribute_fee_to_partner::<T>(who, fee.into())?;
 
+		// if the call is an action permissions transact call, we need to determine
+		// if the spender should be the grantor or grantee
+		if let Some(grantor) = is_action_permission_execute_call::<T>(call) {
+			pallet_sylo_action_permissions::TransactPermissions::<T>::try_mutate(
```
Comment (surangap) [inline pallet/fee-proxy/src/impls.rs:101]: what do you think about using a helper function in the pallet to mutate the storage there rather than directly doing it in the runtime? the same function can return the account that's going to pay the gas and update it in the runtime accordingly.

File: pallet/sylo-action-permissions/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,518 @@
...
+		Blake2_128Concat,
+		T::AccountId, // Grantor
+		Blake2_128Concat,
+		T::AccountId, // Grantee
+		TransactPermission<BlockNumberFor<T>, T::MaxCallIds, T::StringLimit>,
+		OptionQuery,
+	>;
+
+	#[pallet::storage]
+	pub type TokenSignatureNonces<T: Config> = StorageMap<
```
Comment (surangap) [inline pallet/sylo-action-permissions/src/lib.rs:152]: https://github.com/futureversecom/trn-seed/pull/950/files#r2144297998

File: pallet/sylo-action-permissions/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,518 @@
...
+			}
+
+			// Ensure the origin is the grantee
+			ensure!(grantee == permission_token.grantee, Error::<T>::GranteeDoesNotMatch);
+
+			// Validate the nonce
+			ensure!(
+				!TokenSignatureNonces::<T>::contains_key(permission_token.nonce),
+				Error::<T>::NonceAlreadyUsed
+			);
```
Comment (surangap) [inline pallet/sylo-action-permissions/src/lib.rs:361]: I guess this is not the account nonce of the grantee?. How should the nonce be chosen when constructing `permission_token` ?

File: pallet/sylo-action-permissions/src/tests.rs [hunk 4]
```diff
@@ -0,0 +1,917 @@
...
+
+use std::collections::BTreeSet;
+
+use super::*;
+use frame_support::{assert_noop, assert_ok};
+use mock::*;
+use seed_pallet_common::test_prelude::*;
+use sp_core::U256;
+
+fn to_call_id(method: &str, extrinsic: &str) -> CallId<<Test as Config>::StringLimit> {
```
Comment (surangap) [inline pallet/sylo-action-permissions/src/tests.rs:24]: nit
```suggestion
fn to_call_id(pallet_name: &str, extrinsic: &str) -> CallId<<Test as Config>::StringLimit> {
```

File: pallet/sylo-action-permissions/src/lib.rs (hunk 3 above)
Comment (JCSanPedro) [reply pallet/sylo-action-permissions/src/lib.rs:361]: No the client randomly generates a 32 byte value locally. This should be large enough to avoid any conflicts.

Comment (JCSanPedro) [reply pallet/sylo-action-permissions/src/lib.rs:361]: No the client randomly generates a 32 byte value locally. This should be large enough to avoid any conflicts.

File: pallet/sylo-action-permissions/src/types.rs [hunk 5]
```diff
@@ -0,0 +1,270 @@
...
+#[derive(Encode, Decode, TypeInfo, Copy, MaxEncodedLen, Debug, Clone, PartialEq, Eq)]
+pub enum Spender {
+	GRANTOR,
+	GRANTEE,
+}
+
+// Tuple of pallet name and extrinsic name which identifies a specific runtime call.
+pub type CallId<StringLimit> = (BoundedVec<u8, StringLimit>, BoundedVec<u8, StringLimit>);
+
+pub fn to_call_id<StringLimit>(pallet: &str, function: &str) -> CallId<StringLimit>
```
[the remark this answers is not in the document]
Comment (JasonTulp) [reply pallet/sylo-action-permissions/src/types.rs:43]: Hmm interesting, ok not worth the hassle :)

File: pallet/sylo-action-permissions/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,518 @@
...
+
+		/// A lookup to get the futurepass account id for a futurepass holder.
+		type FuturepassLookup: StaticLookup<Source = H160, Target = H160>;
+	}
+
+	/// A set of calls that will always be blocked from being used in a transact
+	/// call. In general we forbid any calls that can be used to submit an
+	/// extrinsic as another account.
+	#[pallet::type_value]
+	pub fn BlacklistedCalls<T: Config>() -> BoundedVec<CallId<T::StringLimit>, T::MaxCallIds>
```
[the remark this answers is not in the document]
Comment (JCSanPedro) [reply pallet/sylo-action-permissions/src/lib.rs:102]: Hmm good catch. I think the only way would be to recursively unwrap the calls when checking if the call is blacklisted.

The approach isn't ideal since if we add any more extrinsics that wrap calls, we would have to make sure update the logic here as well. But not sure of another way to do it.

[the remark this answers is not in the document]
Comment (JasonTulp) [reply pallet/sylo-action-permissions/src/lib.rs:102]: Yeah it gets messy.. What problem is this trying to solve? It may not be required? 
- Sudo would never be possible anyway (Unless Grantor is sudo) 
- Futurepass would theoretically work if the EOA account granted full permissions to an account, but then I feel like that's the risk they take? (Same for proxy)

Unless there are requirements otherwise it could be acceptable to document the risk of full permissions meaning FULL permissions, including any attached Futurepass addresses and highlight the warnings of providing such access

**Review by surangap — COMMENTED** [review]
Nice work 👏. added some comments.

File: pallet/sylo-action-permissions/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,518 @@
...
+		/// The grantee in the token does not match the caller.
+		GranteeDoesNotMatch,
+		/// The nonce provided in the token has already been used.
+		NonceAlreadyUsed,
+		/// The futurepass in the token is not owned by the grantor.
+		InvalidFuturepassInToken,
+	}
+
+	#[pallet::storage]
+	pub type TransactPermissions<T: Config> = StorageDoubleMap<
```
[the remark this answers is not in the document]
Comment (JCSanPedro) [reply pallet/sylo-action-permissions/src/lib.rs:137]: https://github.com/futureversecom/trn-seed/pull/950/commits/3b6368da91cd677bcf67215760ea2727d63ddc5d

File: pallet/sylo-action-permissions/src/tests.rs [hunk 8]
```diff
@@ -0,0 +1,917 @@
...
+					allowed_calls: new_allowed_calls.into_iter().collect(),
+					expiry: new_expiry,
+				}
+				.into(),
+			);
+		});
+	}
+
+	#[test]
+	fn test_update_transact_permission_not_granted() {
```
[the remark this answers is not in the document]
Comment (JCSanPedro) [reply pallet/sylo-action-permissions/src/tests.rs:524]: https://github.com/futureversecom/trn-seed/pull/950/commits/59e425966206f92c2aa84c72a94f5037988f1adc

File: pallet/sylo-action-permissions/src/lib.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (JCSanPedro) [reply pallet/sylo-action-permissions/src/lib.rs]: I had a go at solving this here https://github.com/futureversecom/trn-seed/pull/950/commits/9590781ff81d841a65bee687fa34d050cb627dd2. I think we only need to be concerned about the various utility calls.

**Review by surangap — APPROVED** [review]
lgtm

**Review by JasonTulp — APPROVED** [review]
lgtm! Nice work man, I may have missed them but I think there were two unit tests missing. Otherwise good to go

File: pallet/sylo-action-permissions/src/lib.rs [hunk 9]
```diff
@@ -0,0 +1,561 @@
...
+				T::XrplMaxMessageLength,
+				T::XrplMaxSignatureLength,
+			>,
+		) -> DispatchResult {
+			let grantee = ensure_signed(origin)?;
+
+			// Verify the signature
+			let token_signer = token_signature
+				.verify_signature(&permission_token)
+				.map_err(|_| Error::<T>::InvalidTokenSignature)?;
```
[the remark this answers is not in the document]
Comment (JCSanPedro) [reply pallet/sylo-action-permissions/src/lib.rs:352]: https://github.com/futureversecom/trn-seed/pull/950/commits/f1fed908e90e2976e78f2195134d0ad66535a393

**Review by JasonTulp — APPROVED** [review]
lgtm!
